src/Core/Mesh/TriangleMesh.hpp

```diff
@@ -19,67 +17,89 @@ namespace Core
+    /// Note: Directly using the AttribManager is highly discouraged.
     TriangleMesh( const TriangleMesh& ) = default;
     TriangleMesh& operator=( const TriangleMesh& ) = default;
 
+    /// Copy only the required attributes (deep copy).
+    /// The position and normal attributes are always copied, no need to provide their handles.
+    /// Note: The original handles are not valid for the mesh copy.
+    template <typename... Handles>
+    void partialCopy( const TriangleMesh& input, Handles... attribs );
+
+    /// Copy all the attributes (deep copy).
+    /// Note: The original handles are also valid for the mesh copy.
+    inline void fullCopy( const TriangleMesh& input );
+
     /// Erases all data, making the mesh empty.
+    /// Note: invalidates shallow copies.
     inline void clear();
 
     /// Appends another mesh to this one.
     /// \todo handle attrib here as well !
     inline void append( const TriangleMesh& other );
 
-    VectorArray<Triangle> m_triangles;
-    VectorArray<Face> m_faces;
-
-    PointAttribHandle::Container& vertices()
-    {
+    /// Access the vertices positions.
+    PointAttribHandle::Container& vertices() {
         return m_vertexAttribs.getAttrib( m_verticesHandle ).data();
     }
-    NormalAttribHandle::Container& normals()
-    {
-        return m_vertexAttribs.getAttrib( m_normalsHandle ).data();
+    const PointAttribHandle::Container& vertices() const {
+        return m_vertexAttribs.getAttrib( m_verticesHandle ).data();
     }
 
-    const PointAttribHandle::Container& vertices() const
-    {
-        return m_vertexAttribs.getAttrib( m_verticesHandle ).data();
+    /// Access the vertices normals.
+    NormalAttribHandle::Container& normals() {
+        return m_vertexAttribs.getAttrib( m_normalsHandle ).data();
     }
-    const NormalAttribHandle::Container& normals() const
-    {
+    const NormalAttribHandle::Container& normals() const {
         return m_vertexAttribs.getAttrib( m_normalsHandle ).data();
     }
 
+    /// Access the attribute manager
     const AttribManager& attribManager() const { return m_vertexAttribs; }
     AttribManager& attribManager() { return m_vertexAttribs; }
 
+  public:
+    /// The list of triangles.
+    VectorArray<Triangle> m_triangles;
+
+    /// A list of non-triangular polygons.
+    // FIXME: not used.
+    VectorArray<Face> m_faces;
+
   private:
+    /// The attrib manager.
     AttribManager m_vertexAttribs;
+
+    /// The handle for positions, making request faster.
     PointAttribHandle m_verticesHandle;
+
+    /// The handle for normals, making request faster.
     NormalAttribHandle m_normalsHandle;
 
-    inline void initDefaultAttribs()
-    {
+    /// Sets the default attribs.
+    inline void initDefaultAttribs() {
         m_verticesHandle =
             m_vertexAttribs.addAttrib<PointAttribHandle::value_type>( "in_position" );
         m_normalsHandle = m_vertexAttribs.addAttrib<NormalAttribHandle::value_type>( "in_normal" );
```

Comment:
Do we always have normals ?

Reply:
yes
if not, there are set to 0 ?
but normals are initialized everywhere as a always present attrib